File: kubarr/api/routers/roles.py
"""Role management API endpoints."""

from datetime import datetime
from typing import List, Optional

from fastapi import APIRouter, Depends, HTTPException, status
from pydantic import BaseModel
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import selectinload

from kubarr.api.dependencies import (
    get_current_active_user,
    get_current_admin_user,
    get_db,
)
from kubarr.core.models_auth import Role, RoleAppPermission, User

router = APIRouter()


# Pydantic schemas
class RoleAppPermissionResponse(BaseModel):
    id: int
    app_name: str

    class Config:
        from_attributes = True


class RoleBase(BaseModel):
    name: str
    description: Optional[str] = None


class RoleCreate(RoleBase):
    app_names: List[str] = []


class RoleUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None


class RoleResponse(RoleBase):
    id: int
    is_system: bool
    created_at: datetime
    app_permissions: List[RoleAppPermissionResponse] = []

    class Config:
        from_attributes = True


class RoleWithAppsResponse(BaseModel):
    id: int
    name: str
    description: Optional[str]
    is_system: bool
    created_at: datetime
    app_names: List[str]

    class Config:
        from_attributes = True


class SetRoleAppsRequest(BaseModel):
    app_names: List[str]


@router.get("/", response_model=List[RoleWithAppsResponse])
async def list_roles(
    current_user: User = Depends(get_current_active_user),
    db: AsyncSession = Depends(get_db)
):
    """List all roles.

    Args:
        current_user: Current authenticated user
        db: Database session

    Returns:
        List of roles with their app permissions
    """
    result = await db.execute(
        select(Role).options(selectinload(Role.app_permissions))
    )
    roles = result.scalars().all()

    return [
        RoleWithAppsResponse(
            id=role.id,
            name=role.name,
            description=role.description,
            is_system=role.is_system,
            created_at=role.created_at,
            app_names=[perm.app_name for perm in role.app_permissions]
        )
        for role in roles
    ]


@router.get("/{role_id}", response_model=RoleWithAppsResponse)
async def get_role(
    role_id: int,
    current_user: User = Depends(get_current_active_user),
    db: AsyncSession = Depends(get_db)
):
    """Get role by ID.

    Args:
        role_id: Role ID
        current_user: Current authenticated user
        db: Database session

    Returns:
        Role with app permissions

    Raises:
        HTTPException: If role not found
    """
    result = await db.execute(
        select(Role)
        .options(selectinload(Role.app_permissions))
        .where(Role.id == role_id)
    )
    role = result.scalar_one_or_none()

    if not role:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Role not found"
        )

    return RoleWithAppsResponse(
        id=role.id,
        name=role.name,
        description=role.description,
        is_system=role.is_system,
        created_at=role.created_at,
        app_names=[perm.app_name for perm in role.app_permissions]
    )


@router.post("/", response_model=RoleWithAppsResponse, status_code=status.HTTP_201_CREATED)
async def create_role(
    role_data: RoleCreate,
    current_user: User = Depends(get_current_admin_user),
    db: AsyncSession = Depends(get_db)
):
    """Create a new role (admin only).

    Args:
        role_data: Role creation data
        current_user: Current admin user
        db: Database session

    Returns:
        Created role

    Raises:
        HTTPException: If role name already exists
    """
    # Check if role name exists
    result = await db.execute(
        select(Role).where(Role.name == role_data.name)
    )
    if result.scalar_one_or_none():
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Role name already exists"
        )

    # Create role
    new_role = Role(
        name=role_data.name,
        description=role_data.description,
        is_system=False  # User-created roles are not system roles
    )
    db.add(new_role)
    await db.flush()

    # Add app permissions
    for app_name in role_data.app_names:
        perm = RoleAppPermission(role_id=new_role.id, app_name=app_name)
        db.add(perm)

    await db.commit()
    await db.refresh(new_role)

    # Reload with permissions
    result = await db.execute(
        select(Role)
        .options(selectinload(Role.app_permissions))
        .where(Role.id == new_role.id)
    )
    role = result.scalar_one()

    return RoleWithAppsResponse(
        id=role.id,
        name=role.name,
        description=role.description,
        is_system=role.is_system,
        created_at=role.created_at,
        app_names=[perm.app_name for perm in role.app_permissions]
    )


@router.patch("/{role_id}", response_model=RoleWithAppsResponse)
async def update_role(
    role_id: int,
    role_data: RoleUpdate,
    current_user: User = Depends(get_current_admin_user),
    db: AsyncSession = Depends(get_db)
):
    """Update role (admin only).

    Args:
        role_id: Role ID
        role_data: Role update data
        current_user: Current admin user
        db: Database session

    Returns:
        Updated role

    Raises:
        HTTPException: If role not found or trying to rename system role
    """
    result = await db.execute(
        select(Role)
        .options(selectinload(Role.app_permissions))
        .where(Role.id == role_id)
    )
    role = result.scalar_one_or_none()

    if not role:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Role not found"
        )

    # Prevent renaming system roles
    if role.is_system and role_data.name is not None and role_data.name != role.name:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Cannot rename system roles"
        )

    # Check for duplicate name
    if role_data.name is not None and role_data.name != role.name:
        existing = await db.execute(
            select(Role).where(Role.name == role_data.name)
        )
        if existing.scalar_one_or_none():
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Role name already exists"
            )

    # Update fields
    if role_data.name is not None:
        role.name = role_data.name
    if role_data.description is not None:
        role.description = role_data.description

    await db.commit()
    await db.refresh(role)

    return RoleWithAppsResponse(
        id=role.id,
        name=role.name,
        description=role.description,
        is_system=role.is_system,
        created_at=role.created_at,
        app_names=[perm.app_name for perm in role.app_permissions]
    )


@router.delete("/{role_id}")
async def delete_role(
    role_id: int,
    current_user: User = Depends(get_current_admin_user),
    db: AsyncSession = Depends(get_db)
):
    """Delete a role (admin only).

    System roles cannot be deleted.

    Args:
        role_id: Role ID
        current_user: Current admin user
        db: Database session

    Returns:
        Success message

    Raises:
        HTTPException: If role not found or is a system role
    """
    result = await db.execute(
        select(Role).where(Role.id == role_id)
    )
    role = result.scalar_one_or_none()

    if not role:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Role not found"
        )

    if role.is_system:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Cannot delete system roles"
        )

    await db.delete(role)
    await db.commit()

    return {"message": "Role deleted"}


@router.put("/{role_id}/apps", response_model=RoleWithAppsResponse)
async def set_role_apps(
    role_id: int,
    apps_data: SetRoleAppsRequest,
    current_user: User = Depends(get_current_admin_user),
    db: AsyncSession = Depends(get_db)
):
    """Set app permissions for a role (admin only).

    This replaces all existing app permissions for the role.

    Args:
        role_id: Role ID
        apps_data: List of app names
        current_user: Current admin user
        db: Database session

    Returns:
        Updated role with new permissions

    Raises:
        HTTPException: If role not found
    """
    result = await db.execute(
        select(Role)
        .options(selectinload(Role.app_permissions))
        .where(Role.id == role_id)
    )
    role = result.scalar_one_or_none()

    if not role:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Role not found"
        )

    # Delete existing permissions
    for perm in role.app_permissions:
        await db.delete(perm)

    # Add new permissions
    for app_name in apps_data.app_names:
        perm = RoleAppPermission(role_id=role.id, app_name=app_name)
        db.add(perm)

    await db.commit()

    # Reload with new permissions
    result = await db.execute(
        select(Role)
        .options(selectinload(Role.app_permissions))
        .where(Role.id == role_id)
    )
    role = result.scalar_one()

    return RoleWithAppsResponse(
        id=role.id,
        name=role.name,
        description=role.description,
        is_system=role.is_system,
        created_at=role.created_at,
        app_names=[perm.app_name for perm in role.app_permissions]
    )

File: kubarr-rs/src/api/roles.rs
use axum::{
    extract::{Path, State},
    routing::{delete, get, patch, post, put},
    Json, Router,
};
use serde::{Deserialize, Serialize};

use crate::api::extractors::{AdminUser, AuthUser};
use crate::db::{DbPool, Role, RoleAppPermission};
use crate::error::{AppError, Result};
use crate::state::AppState;

/// Create roles routes
pub fn roles_routes(state: AppState) -> Router {
    Router::new()
        .route("/", get(list_roles).post(create_role))
        .route("/:role_id", get(get_role).patch(update_role).delete(delete_role))
        .route("/:role_id/apps", put(set_role_apps))
        .with_state(state)
}

// ============================================================================
// Request/Response Types
// ============================================================================

#[derive(Debug, Deserialize)]
pub struct CreateRole {
    pub name: String,
    pub description: Option<String>,
    #[serde(default)]
    pub app_names: Vec<String>,
}

#[derive(Debug, Deserialize)]
pub struct UpdateRole {
    pub name: Option<String>,
    pub description: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct SetRoleApps {
    pub app_names: Vec<String>,
}

#[derive(Debug, Serialize)]
pub struct RoleWithAppsResponse {
    pub id: i64,
    pub name: String,
    pub description: Option<String>,
    pub is_system: bool,
    pub created_at: chrono::DateTime<chrono::Utc>,
    pub app_names: Vec<String>,
}

// ============================================================================
// Helper Functions
// ============================================================================

async fn get_role_with_apps(pool: &DbPool, role_id: i64) -> Result<RoleWithAppsResponse> {
    let role: Role = sqlx::query_as("SELECT * FROM roles WHERE id = ?")
        .bind(role_id)
        .fetch_optional(pool)
        .await?
        .ok_or_else(|| AppError::NotFound("Role not found".to_string()))?;

    let permissions: Vec<RoleAppPermission> =
        sqlx::query_as("SELECT * FROM role_app_permissions WHERE role_id = ?")
            .bind(role_id)
            .fetch_all(pool)
            .await?;

    Ok(RoleWithAppsResponse {
        id: role.id,
        name: role.name,
        description: role.description,
        is_system: role.is_system,
        created_at: role.created_at,
        app_names: permissions.into_iter().map(|p| p.app_name).collect(),
    })
}

// ============================================================================
// Endpoint Handlers
// ============================================================================

/// List all roles
async fn list_roles(
    State(state): State<AppState>,
    AuthUser(_): AuthUser,
) -> Result<Json<Vec<RoleWithAppsResponse>>> {
    let roles: Vec<Role> = sqlx::query_as("SELECT * FROM roles")
        .fetch_all(&state.pool)
        .await?;

    let mut responses = Vec::new();
    for role in roles {
        responses.push(get_role_with_apps(&state.pool, role.id).await?);
    }

    Ok(Json(responses))
}

/// Get role by ID
async fn get_role(
    State(state): State<AppState>,
    Path(role_id): Path<i64>,
    AuthUser(_): AuthUser,
) -> Result<Json<RoleWithAppsResponse>> {
    let response = get_role_with_apps(&state.pool, role_id).await?;
    Ok(Json(response))
}

/// Create a new role (admin only)
async fn create_role(
    State(state): State<AppState>,
    AdminUser(_): AdminUser,
    Json(data): Json<CreateRole>,
) -> Result<Json<RoleWithAppsResponse>> {
    // Check if role name exists
    let existing: Option<Role> = sqlx::query_as("SELECT * FROM roles WHERE name = ?")
        .bind(&data.name)
        .fetch_optional(&state.pool)
        .await?;

    if existing.is_some() {
        return Err(AppError::BadRequest("Role name already exists".to_string()));
    }

    // Create role
    let result = sqlx::query(
        r#"
        INSERT INTO roles (name, description, is_system, created_at)
        VALUES (?, ?, 0, datetime('now'))
        "#,
    )
    .bind(&data.name)
    .bind(&data.description)
    .execute(&state.pool)
    .await?;

    let role_id = result.last_insert_rowid();

    // Add app permissions
    for app_name in &data.app_names {
        sqlx::query("INSERT INTO role_app_permissions (role_id, app_name) VALUES (?, ?)")
            .bind(role_id)
            .bind(app_name)
            .execute(&state.pool)
            .await?;
    }

    let response = get_role_with_apps(&state.pool, role_id).await?;
    Ok(Json(response))
}

/// Update role (admin only)
async fn update_role(
    State(state): State<AppState>,
    Path(role_id): Path<i64>,
    AdminUser(_): AdminUser,
    Json(data): Json<UpdateRole>,
) -> Result<Json<RoleWithAppsResponse>> {
    let role: Role = sqlx::query_as("SELECT * FROM roles WHERE id = ?")
        .bind(role_id)
        .fetch_optional(&state.pool)
        .await?
        .ok_or_else(|| AppError::NotFound("Role not found".to_string()))?;

    // Prevent renaming system roles
    if role.is_system && data.name.is_some() && data.name.as_ref() != Some(&role.name) {
        return Err(AppError::BadRequest(
            "Cannot rename system roles".to_string(),
        ));
    }

    // Check for duplicate name
    if let Some(ref new_name) = data.name {
        if new_name != &role.name {
            let existing: Option<Role> = sqlx::query_as("SELECT * FROM roles WHERE name = ?")
                .bind(new_name)
                .fetch_optional(&state.pool)
                .await?;

            if existing.is_some() {
                return Err(AppError::BadRequest("Role name already exists".to_string()));
            }
        }
    }

    // Update fields
    if let Some(name) = &data.name {
        sqlx::query("UPDATE roles SET name = ? WHERE id = ?")
            .bind(name)
            .bind(role_id)
            .execute(&state.pool)
            .await?;
    }

    if let Some(description) = &data.description {
        sqlx::query("UPDATE roles SET description = ? WHERE id = ?")
            .bind(description)
            .bind(role_id)
            .execute(&state.pool)
            .await?;
    }

    let response = get_role_with_apps(&state.pool, role_id).await?;
    Ok(Json(response))
}

/// Delete a role (admin only)
async fn delete_role(
    State(state): State<AppState>,
    Path(role_id): Path<i64>,
    AdminUser(_): AdminUser,
) -> Result<Json<serde_json::Value>> {
    let role: Role = sqlx::query_as("SELECT * FROM roles WHERE id = ?")
        .bind(role_id)
        .fetch_optional(&state.pool)
        .await?
        .ok_or_else(|| AppError::NotFound("Role not found".to_string()))?;

    if role.is_system {
        return Err(AppError::BadRequest(
            "Cannot delete system roles".to_string(),
        ));
    }

    sqlx::query("DELETE FROM roles WHERE id = ?")
        .bind(role_id)
        .execute(&state.pool)
        .await?;

    Ok(Json(serde_json::json!({"message": "Role deleted"})))
}

/// Set app permissions for a role (admin only)
async fn set_role_apps(
    State(state): State<AppState>,
    Path(role_id): Path<i64>,
    AdminUser(_): AdminUser,
    Json(data): Json<SetRoleApps>,
) -> Result<Json<RoleWithAppsResponse>> {
    let _: Role = sqlx::query_as("SELECT * FROM roles WHERE id = ?")
        .bind(role_id)
        .fetch_optional(&state.pool)
        .await?
        .ok_or_else(|| AppError::NotFound("Role not found".to_string()))?;

    // Delete existing permissions
    sqlx::query("DELETE FROM role_app_permissions WHERE role_id = ?")
        .bind(role_id)
        .execute(&state.pool)
        .await?;

    // Add new permissions
    for app_name in &data.app_names {
        sqlx::query("INSERT INTO role_app_permissions (role_id, app_name) VALUES (?, ?)")
            .bind(role_id)
            .bind(app_name)
            .execute(&state.pool)
            .await?;
    }

    let response = get_role_with_apps(&state.pool, role_id).await?;
    Ok(Json(response))
}
